### ISSUE TITLE
Angular 2 and dx.aspnet.data.d.ts

### ISSUE THREAD

**Issue by Dmitry44** [DevExpress/DevExtreme.AspNet.Data#91]
Hi!
I am tring to use dx.aspnet.data.js from typescript (Angular 2).
As far as I understand it is needed to create interface definition file - dx.aspnet.data.d.ts
Maybe somebody did it? Or there is some example of using devextreme's aspnet data store from Angular 2?

**Comment by AlekseyMartynov** [comment]
@Dmitry44
Aside from a definition file, the script needs to be AMD compliant.
We'll fix it, prepare an example and update this thread.

**Comment by AlekseyMartynov** [comment]
@GoshaFighten version 1.2.2 is published

**Comment by AlekseyMartynov** [comment]
@Dmitry44 
Starting with version 1.2.3, the script is available [via NPM](https://www.npmjs.com/package/devextreme-aspnet-data).

`npm i --save devextreme-aspnet-data`

Use it in your Angular app as follows:

**TS:**

```js
import * as AspNetData from "devextreme-aspnet-data";

// . . .
export class HomeComponent implements OnInit {
    // . . .
    store = AspNetData.createStore({
        key: "OrderID",
        loadUrl: "api/SampleData/orders"
    });
}
```

**HTML:**

```html
<dx-data-grid [dataSource]="store" [remoteOperations]="true"></dx-data-grid>
```

**Comment by Dmitry44** [comment]
@AlekseyMartynov  Hi and thank you!
I tried to use 1.2.2 but didn't fix some problems. Can you share your devextreme-aspnet-data.d.ts (index.d.ts) file?

**Comment by AlekseyMartynov** [comment]
Yes, 1.2.2 doesn't contain all changes necessary for Angular compatibility.

Please try the following:
1. Uninstall bower package
2. Install 1.2.3 from npm: https://www.npmjs.com/package/devextreme-aspnet-data
3. Import the lib as shown in my [previous comment](https://github.com/DevExpress/DevExtreme.AspNet.Data/issues/91#issuecomment-273107522)

We checked that both Webpack and TypeScript compiler work fine without d.ts.

**Comment by Dmitry44** [comment]
@AlekseyMartynov 
Visual Studio 2015 shows error:
![image](https://cloud.githubusercontent.com/assets/3319327/22070628/c75749d8-ddad-11e6-859b-046f3826d7bd.png)

But you are right. It works despite of this!

**Comment by GoshaFighten** [comment]
From GitHub
 
@Dmitry44 
To be honest, I'm not sure why Visual Studio can't find our "devextreme-aspnet-data" module. But, you can require it using the following code:
``` TypeScript
let AspNetData = require('devextreme-aspnet-data');
```
If you are still unable to run your project, open a new ticket in our [Support Center](https://www.devexpress.com/Support/Center/Question/Create).

**Comment by AlekseyMartynov** [comment]
As I understand, there's no problem to build and run. The error appears only in Visual Studio code editor.

@kvet Please take a look. Is it the same issue due to which we need per-module definition files in DevExtreme npm?

**Comment by kvet** [comment]
First of all, the `devextreme-aspnet-data` package doesn't have definitions, which causes building errors in TS.
 
As a workaround, I can suggest adding an ambient module declaration for the `devextreme-aspnet-data` package described in the "Working with Other JavaScript Libraries" section of [TS Modules Guide](https://www.typescriptlang.org/docs/handbook/modules.html).
Also, note that TS 2.1 no longer requires such a module declaration. See the "Untyped imports" section of [TS 2.1 Release Notes](https://www.typescriptlang.org/docs/handbook/release-notes/typescript-2-1.html).
 
To fix this issue "completely", we should add built-in package typings. Or, as an alternative solution, we could publish typings to the DefinitelyTyped repo.

**Comment by AlekseyMartynov** [comment]
@kvet
Just to make sure I understand correctly:
1. Definitions are required only for old versions of TS (< 2.1)
2. It's sufficient to add a minimal ambient d.ts, like a function accepting `any` and returning `any`. 

P.S. My motivation regarding point 2 is that it seems too much to have a full definition with deps on jQuery and DevExtreme for such a small piece of API.

**Comment by kvet** [comment]
@AlekseyMartynov 
1. Yes.
2. Following ambient declaration should be enough to prevent build errors:
```ts
/* ambient-declarations.d.ts */

declare module "devextreme-aspnet-data";
```

**Comment by Dmitry44** [comment]
> With TypeScript 2.1, you can import a JavaScript module without needing a type declaration. A type declaration (such as declare module "foo" { ... } or node_modules/@types/foo) still takes priority if it exists.
> An import to a module with no declaration file will still be flagged as an error under --noImplicitAny.
> Example
> // Succeeds if `node_modules/asdf/index.js` exists
> import { x } from "asdf";

I've updated typescript to the version 2.1.5. The error exists still. I didn't find --noImplicitAny option in my configuration files. Looks like Visual Studio doesn't support this future yet

**Comment by AlekseyMartynov** [comment]
@Dmitry44 
Have you also updated Visual Studio extension?
https://www.microsoft.com/en-us/download/details.aspx?id=48593

**Comment by Dmitry44** [comment]
Hi @AlekseyMartynov 
Yes, I did
Visual Sudio extension and nodejs package were updated to 2.1.5

**Comment by AlekseyMartynov** [comment]
Thanks. 
I prepared a PR #98 
@Dmitry44 may I ask you to check the intermediate npm build available at https://ci.appveyor.com/project/dxrobot/devextreme-aspnet-data/build/233/artifacts ?

Install it with the following command:

`npm install devextreme-aspnet-data-99.0.0-ci-000233.tgz`

**Comment by Dmitry44** [comment]
@AlekseyMartynov 
Sorry for delay - long weekend
There is error during install, but package was installed:

> npm install devextreme-aspnet-data-99.0.0-ci-000233.tgz
> `-- devextreme-aspnet-data@99.0.0-ci-000233  invalid

There is no error in VS now.
But another error was raised when I tried to sort my grid:
> (18,54): error TS2304: Cannot find name 'JQueryAjaxSettings'

It was fixed by using any instead of JQueryAjaxSettings:

> onBeforeSend?: (operation: string, ajaxSettings: any) => void

**Comment by AlekseyMartynov** [comment]
@Dmitry44 

> There is error during install, but package was installed

Our of curiosity, what is the output of `npm -v`?

> (18,54): error TS2304: Cannot find name 'JQueryAjaxSettings'

Adding [@types/jquery](https://www.npmjs.com/package/@types/jquery) package should fix this error. We [discussed it](https://github.com/DevExpress/DevExtreme.AspNet.Data/pull/98/files/2ac8216265f37c9cf5a000fec7c49e7be3c75873#diff-b9cfc7f2cdf78a7f4b91a753d10865a2R18) with @kvet and decided that we shouldn't add the dependency. By the way, @types/jquery should have beed installed as a dependency of [devextreme-angular](https://www.npmjs.com/package/devextreme-angular). Could you please check again? Maybe on a clean project.

**Comment by Dmitry44** [comment]
@AlekseyMartynov 

@types/jquery package is installed.

> npm -v
> 3.10.9

After thiis:
- npm uninstall devextreme-aspnet-data
- npm install npm -g (4.1.1 was installed)
- npm install devextreme-aspnet-data-99.0.0-ci-000233.tgz

The same error: "Cannot find name 'JQueryAjaxSettings'"

![image](https://cloud.githubusercontent.com/assets/3319327/22202913/75de63bc-e17b-11e6-9ba1-367b3f13c5b9.png)

**Comment by AlekseyMartynov** [comment]
@kvet thoughts?

**Comment by kvet** [comment]
@Dmitry44 have you tried to specify `typesRoot` option for `tsconfig.json`? It should point to `node_modules/@types` folder in your case.

**Comment by Dmitry44** [comment]
@AlekseyMartynov @kvet 
tsconfig.json documentation said:
> By default all visible “@types” packages are included in your compilation. Packages in node_modules/@types of any enclosing folder are considered visible

But I found that node_modules/@types/jquery doesn't included in my case.
This tsconfig.json works for me:

```
{
  "compilerOptions": {
    "moduleResolution": "node",
    "target": "es5",
    "sourceMap": true,
    "experimentalDecorators": true,
    "emitDecoratorMetadata": true,
    "skipDefaultLibCheck": true,
    "lib": [ "es6", "dom" ],
    "types": [ "node", "jquery" ]
  },
  "exclude": [ "bin", "node_modules" ],
  "atom": { "rewriteTsconfig": false }
}
```
It is not convinient to include types explicitly but it works.
Anyway thank you very much!

**Comment by AlekseyMartynov** [comment]
Good.
The changes will be included in 1.2.4

**Comment by dinh383** [comment]
Hey, everybody.
I was find example about Data Grid of Devextreme with angular 2, processing insert, update, delete on server side use library npm install devextreme-aspnet-data but i don't find source code example. Can you send for me source code angular 2 with data grid call API processing?
Thanks you all.
